src/Database/LSMTree/Internal/MergingRun.hs

```diff
@@ -211,265 +226,502 @@ duplicateRuns (DeRef mr) =
+                performMergeSteps mergeState mergeCreditsVar spendCredits
+
+              -- If an async exception happens before we get to perform the
+              -- completion, then that is fine. The next supplyCredits will
+              -- complete the merge.
+              when weFinishedMerge $
+                completeMerge mergeState mergeKnownCompleted
+
+            return leftoverCredits)
+
+{-# SPECIALISE performMergeSteps ::
      StrictMVar IO (MergingRunState IO h)
-  -> TotalStepsVar RealWorld
+  -> CreditsVar RealWorld
   -> Credits
   -> IO Bool #-}
-stepMerge ::
+performMergeSteps ::
      (MonadMVar m, MonadMask m, MonadSTM m, MonadST m)
   => StrictMVar m (MergingRunState m h)
-  -> TotalStepsVar (PrimState m)
+  -> CreditsVar (PrimState m)
   -> Credits
   -> m Bool
-stepMerge mergeVar (TotalStepsVar totalStepsVar) (Credits c) =
+performMergeSteps mergeVar creditsVar (Credits credits) =
     withMVar mergeVar $ \case
-      CompletedMerge{} -> pure False
-      (OngoingMerge _rs (SpentCreditsVar spentCreditsVar) m) -> do
-        totalSteps <- readPrimVar totalStepsVar
-        spentCredits <- readPrimVar spentCreditsVar
-
-        -- If we previously performed too many merge steps, then we perform
-        -- fewer now.
-        let stepsToDo = max 0 (spentCredits + c - totalSteps)
-        -- Merge.steps guarantees that @stepsDone >= stepsToDo@ /unless/ the
-        -- merge was just now finished.
+      CompletedMerge{}   -> pure False
+      OngoingMerge _rs m -> do
+        -- We have dealt with the case of credits <= 0 above,
+        -- so here we know credits is positive
+        let stepsToDo = credits
         (stepsDone, stepResult) <- Merge.steps m stepsToDo
-        assert (case stepResult of
-                  MergeInProgress -> stepsDone >= stepsToDo
-                  MergeDone       -> True
-                ) $ pure ()
-
-        -- This should be the only point at which we write to these variables.
-        --
-        -- It is guaranteed that @totalSteps' >= spentCredits'@ /unless/ the
-        -- merge was just now finished.
-        let totalSteps' = totalSteps + stepsDone
-        let spentCredits' = spentCredits + c
-        -- It is guaranteed that
-        -- @readPrimVar totalStepsVar >= readPrimVar spentCreditsVar@,
-        -- /unless/ the merge was just now finished.
-        writePrimVar totalStepsVar $! totalSteps'
-        writePrimVar spentCreditsVar $! spentCredits'
-        assert (case stepResult of
-                  MergeInProgress -> totalSteps' >= spentCredits'
-                  MergeDone       -> True
-              ) $ pure ()
+        assert (stepsDone >= stepsToDo) (pure ())
+        -- Merge.steps guarantees that @stepsDone >= stepsToDo@ /unless/ the
+        -- merge was just now finished and excess credit was supplied. But
+        -- elsewhere we guarantee that we do not supply credit in excess of
+        -- the total debt, so the inequality should hold here always.
```

Comment:
Thinking about this again, I don't think it's right. The steps performed might be larger than the credits used for that, mainly because we put back credits if an exception occurred in the middle of performing merge steps. Therefore, the steps that still need to be performed might be fewer than the credits used here.

Reply:
Mmm. Ok, in which case we would still need the more complex assertion we had before which made it conditional on not having finished (stepResult /= MergeDone).

Reply:
Yes, I think so.

Reply:
Good catch. Fixed.